feat: extension erc 721

tu11aa opened the pull request:
## Feature:
- Implement constructor with metadata and token uri
- New erc721 UI

## Fix:
- change opt-level to 'z' to reduce contract size

tu11aa committed:
WIP: erc 721 with constructor

tu11aa committed:
feat: erc721 contract
fix: change opt-level to 3 to reduce contract size

tu11aa committed:
feat: erc721 new UI

gianalarcon commented on extensions/packages/stylus/erc721-example/src/lib.rs:
```diff
@@ -17,210 +19,242 @@ use stylus_sdk::{
 
-    pub fn safe_mint(&mut self, to: Address, token_id: U256, data: Bytes) -> Result<(), Error> {
-        self.erc721._safe_mint(to, token_id, &data)?;
+    fn total_supply(&self) -> U256 {
```
do we really need to implement fn total_supply() on `Erc721MetadataExample` and also on `IErc721Enumerable`?

gianalarcon replied:
https://github.com/Arb-Stylus/create-stylus-extensions/pull/1/files#diff-d145ff59fab85d335918da265d0d1dafba427bd587fa402845ab53a365216b85

check this config/setup, to fix missing to export abi

tu11aa replied:
- `transfer nft` fixed

> do we really need to implement fn total_supply() on `Erc721MetadataExample` and also on `IErc721Enumerable`?

Yes, must implement like that to export function to abi

> https://github.com/Arb-Stylus/create-stylus-extensions/pull/1/files#diff-d145ff59fab85d335918da265d0d1dafba427bd587fa402845ab53a365216b85
> 
> check this config/setup, to fix missing to export abi

This config is duplicated because stylus-sdk/export-abi will be enabled automatically by this config. I also tried this approach before and it didn't work

gianalarcon reviewed (commented):
Looks good, double check `transfer nft` funcionality

tu11aa committed:
fix: add missing transfer enumeration owner in transfer_from

tu11aa committed:
fix: tests.
WIP: The problem is that openzeppelin-stylus version 0.3.0 is using an older API of stylus-sdk where VM was a struct with a host field, but in stylus-sdk 0.9.0, VM is now a tuple struct
